john's changes w/ cleaned up bindata

borkmann opened the pull request:
(testing whether Jenkins still has issues with that one)

jrfastab commented on tests/14-proxy-ingress.sh:
```diff
@@ -0,0 +1,69 @@
+cilium policy delete --all
+cat <<EOF | cilium -D policy import -
+[{
+    "endpointSelector": ["id.server"],
```
With the recent changes do we need this here, 

``` "endpointSelector": {"matchLabels":{"id.server":""}}, ```

borkmann replied:
Agree, done.

jrfastab commented on tests/14-proxy-ingress.sh:
```diff
@@ -0,0 +1,69 @@
+    "endpointSelector": ["id.server"],
+    "ingress": [{
+        "fromEndpoints": [
+	    ["reserved:host"], ["id.client"]
```
and same here

``` "fromEndpoints": [ {"matchLabels":{"resreved:host":""}},
                                  {"matchLabels":{"id.client":""}}
```

jrfastab reviewed (approved):
The patches look good to me, thanks for fixing this up.

I'll approve these but is there a way to kick jenkins to run the needed tests? I'm having some issues with my local testing, but it appears unrelated to these patches.

jrfastab committed:
cilium: add ingress proxy tests

Add ingress proxy tests to the test infrastructure.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

borkmann committed:
tests, ct: bail out if configuration change caused an error

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

jrfastab committed:
cilium: bindata update due to bpf changes

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: support ingress IPv4 proxy policy

The proxy logic uses the cilium_proxy4 map to lookup the reverse
nat. The key is built as follows,

	struct proxy4_tbl_key key = {
		.saddr = bpf_htonl(LXC_IPV4),
		.sport = tuple->sport,
		.dport = new_port,
		.nexthdr = tuple->nexthdr,
	};

In the egress case ipv4_redirect_to_host_port(), the logic that
populates cilium_proxy4 map and builds the key, is called in the
context of the client's LXC_IPV4. Assuming the typical model,

          client <-> proxy <-> server

So when the proxy replies and a reverse map is needed we have
daddr == LXC_IPV4 and can do a correct map lookup.

However, in the ingress case ipv4_redirect_to_host_port() is
called from the bpf program referenced in the policy map (e.g.
the server in the test cases). In this case LXC_IPV4 is the
IP address of the server. The result is when the proxy replies
and uses the clients IP address in the key we get a miss in the
cilium_proxy4 map.

To resolve the error instead of using the compile-time defined
LXC_IPV4 pull the address out of the tuple data when available
or directly out of the ipv4 header itself. This resolves the
(client <-> proxy) lookups.

On the (proxy <-> server) side the last piece to fixup is to
ensure when the proxy is sending to the server, in the ingress
case, we avoid forwarding to the proxy ports. This happens
because the server has an ingress port map that, unless aborted,
results in header rewrites. Here we add a check to the conntrack
path to catch this case.

With the above fixes both egress and ingress proxy tests are
working again.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>
[ dbkm: merged fix for gw addr test ]
